### PR TITLE
IBP-5031: Implement samples search and refactor samples service

### PR DESCRIPTION
**Pull request by cheryleigh**

### CODE REVIEW

**Review by mderamos — CHANGES_REQUESTED** [review]

File: src/main/java/org/generationcp/middleware/domain/search_request/brapi/v2/SampleSearchRequestDTO.java [hunk 1]
```diff
@@ -1,103 +1,135 @@
...
+	private List<String> externalReferenceSources;
+	private List<String> germplasmNames;
+	private List<String> studyNames;
+
+	public SampleSearchRequestDTO() {
+	}
+
+	public SampleSearchRequestDTO(final String sampleDbIds, final String observationUnitDbIds, final String plateDbIds,
+		final String germplasmDbIds, final String studyDbIds, final String externalReferenceIDs, final String externalReferenceSources) {
+		this.sampleDbIds = Collections.singletonList(sampleDbIds);
```
Comment (mderamos) [inline src/main/java/org/generationcp/middleware/domain/search_request/brapi/v2/SampleSearchRequestDTO.java:28]: perform StringUtils.isEmpty() before putting in singletonList. Also make the parameters singular

**Review by mderamos — APPROVED** [review]

File: src/main/java/org/generationcp/middleware/dao/SampleDao.java [hunk 2]
```diff
@@ -527,29 +534,39 @@ private void appendSamplesFromQuery(final StringBuilder sql) {
...
+		}
+
+		// Search preferred names
+		if (!CollectionUtils.isEmpty(requestDTO.getGermplasmNames())) {
+			sql.append(" AND g.gid IN ( SELECT n.gid ");
+			sql.append(" FROM names n WHERE n.nstat = 1 AND n.nval in (:germplasmNames) ) ");
+		}
+
+		if (!CollectionUtils.isEmpty(requestDTO.getStudyNames())) {
+			sql.append(" AND CONCAT(pmain.name, ' Environment Number ', geoloc.description) in (:studyNames)");
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/dao/SampleDao.java:569]: Can you extract this as constant (maybe `STUDY_NAME_BRAPI`) in `DmsProjectDao` and use it here:
`CONCAT(pmain.name, ' Environment Number ', geoloc.description)`

File: src/main/java/org/generationcp/middleware/service/api/SampleService.java [hunk 3]
```diff
@@ -33,8 +33,4 @@ Sample buildSample(final String cropName, final String cropPrefix, final Integer
 	Boolean studyHasSamples(final Integer studyId);
 
 	Boolean studyEntryHasSamples(final Integer studyId, final Integer entryId);
-
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/service/api/SampleService.java:36]: Remove unused imports

File: src/main/java/org/generationcp/middleware/service/impl/study/SampleServiceImpl.java [hunk 4]
```diff
@@ -128,31 +128,6 @@ public Boolean studyEntryHasSamples(final Integer studyId, final Integer entryId
 		return this.daoFactory.getSampleDao().studyEntryHasSamples(studyId, entryId);
 	}
 
-	@Override
```
Comment (darla-leafnode) [inline src/main/java/org/generationcp/middleware/service/impl/study/SampleServiceImpl.java:131]: Remove unused imports

**Review by darla-leafnode — COMMENTED** [review]
@cheryleigh - please see my review comments

**Review by darla-leafnode — APPROVED** [review]

**Pull request merged**